include/apq/impl/async_connect.h

```diff
@@ -0,0 +1,172 @@
+inline void pq_read_poll(T& conn, Handler&& h) {
+    get_socket(conn).async_read_some(
+            asio::null_buffers(), std::forward<Handler>(h));
+}
+
+template <typename T, typename = Require<Connectiable<T>>>
+inline decltype(auto) pq_connect_poll(T& conn) {
+    return PQconnectPoll(get_native_handle(conn));
+}
+
+template <typename T, typename = Require<Connectiable<T>>>
+inline error_code pq_start_connection(T& conn, const std::string& conninfo) {
+    pg_conn_handle handle(PQconnectStart(conninfo.c_str()), PQfinish);
+    if (!handle) {
+        return make_error_code(error::pq_connection_start_failed);
+    }
+    get_handle(conn) = std::move(handle);
+    return {};
+}
+
+template <typename T, typename = Require<Connectiable<T>>>
+inline error_code pq_assign_socket(T& conn) {
+    int fd = PQsocket(get_native_handle(conn));
+    if (fd == -1) {
+        return make_error_code(error::pq_socket_failed);
+    }
+
+    int new_fd = dup(fd);
+    if (new_fd == -1) {
+        using namespace boost::system;
+        return make_error_code(error_code{errno, get_posix_category()},
+            "dup(fd)");
+    }
+
+    boost::system::error_code ec;
+    get_socket(conn).assign(new_fd, ec);
+
+    return make_error_code(ec, "assign socket failed");
+}
+} // namespace pq
+
+template <typename T, typename = Require<Connectiable<T>>>
+inline error_code start_connection(T& conn, const std::string& conninfo) {
+    using pq::pq_start_connection;
+    return pq_start_connection(unwrap_connection(conn), conninfo);
+}
+
+template <typename T, typename = Require<Connectiable<T>>>
+inline error_code assign_socket(T& conn) {
+    using pq::pq_assign_socket;
+    return pq_assign_socket(unwrap_connection(conn));
+}
+
+template <typename T, typename Handler, typename = Require<Connectiable<T>>>
+inline void write_poll(T& conn, Handler&& h) {
+    using pq::pq_write_poll;
+    pq_write_poll(unwrap_connection(conn), std::forward<Handler>(h));
+}
+
+template <typename T, typename Handler, typename = Require<Connectiable<T>>>
+inline void read_poll(T& conn, Handler&& h) {
+    using pq::pq_read_poll;
+    pq_read_poll(unwrap_connection(conn), std::forward<Handler>(h));
+}
+
+template <typename T, typename = Require<Connectiable<T>>>
+inline decltype(auto) connect_poll(T& conn) {
+    using pq::pq_connect_poll;
+    return pq_connect_poll(unwrap_connection(conn));
+}
+
+/**
+* Asynchronous connection operation
+*/
+template <typename Handler, typename Connection>
+struct async_connect_op {
+    static_assert(Connectiable<Connection>,
+        "Connection type does not meet Connectiable requirements");
+
+    Connection& conn_;
+    Handler handler_;
+
+    void done(error_code ec = error_code{}) {
+        get_io_context(conn_)
+            .post(detail::bind(std::move(handler_), std::move(ec)));
+    }
+
+    void perform(const std::string& conninfo) {
+        error_code ec = start_connection(conn_, conninfo);
+        if (ec) {
```

Comment:
Allow to reuse ec variable is not so good idea. Why just not ```if (const auto ec = start_connection(conn_, conninfo)) {```?

Reply:
**Done**